src/vos/ilog.c

```diff
@@ -1552,3 +1392,204 @@ ilog_fetch_finish(struct ilog_entries *entries)
+
+	/* This can potentially be optimized but using ilog_fetch gets some code
+	 * reuse.
+	 */
+	rc = ilog_fetch(umm, ilog, cbs, DAOS_INTENT_PURGE, entries);
+	if (rc == -DER_NONEXIST) {
+		D_DEBUG(DB_IO, "log is empty\n");
+		/* Log is empty */
+		return 1;
+	}
+
+	lctx = &priv->ip_lctx;
+
+	root = lctx->ic_root;
+
+	ILOG_ASSERT_VALID(root);
+
+	D_ASSERT(!ilog_empty(root)); /* ilog_fetch should have failed */
+
+	if (root->lr_tree.it_embedded) {
+		entry = &entries->ie_entries[0];
+		switch (entry->ie_status) {
+		case ILOG_COMMITTED:
+			if (entry->ie_id.id_epoch < epr->epr_lo ||
+			    entry->ie_id.id_epoch > epr->epr_hi) {
+				rc = 0;
+				goto done;
+			}
+			if (!discard && !entry->ie_punch) {
+				rc = 0;
+				return 0;
+			}
+			/* fallthrough */
+		case ILOG_REMOVED:
+			old_id = root->lr_id;
+			tmp.lr_magic = ilog_ver_inc(lctx);
+			rc = ilog_ptr_set(lctx, root, &tmp);
+			if (rc != 0)
+				break;
+
+			empty = true;
+			rc = ilog_log_del(lctx, &old_id);
+			D_DEBUG(DB_IO, "Removed ilog entry at "DF_U64"\n",
+				entry->ie_id.id_epoch);
+			removed++;
+			break;
+		default:
+			/* Should never be UNCOMMITTED with purge set */
+			D_ASSERT(0);
+		}
+		goto done;
+	}
+
+	umem_attr_get(&lctx->ic_umm, &uma);
+	rc = dbtree_open(root->lr_tree.it_root, &uma, &toh);
+	if (rc != 0) {
+		D_ERROR("Failed to open incarnation log tree: "DF_RC
+			"\n", DP_RC(rc));
+		return rc;
+	}
+	prev = NULL;
+	prior_punch = NULL;
+	ilog_foreach_entry(entries, entry) {
+		D_DEBUG(DB_TRACE, "Entry "DF_U64" punch=%s prev="DF_U64
+			" prior_punch="DF_U64"\n", entry->ie_id.id_epoch,
+			entry->ie_punch ? "yes" : "no",
+			prev ? prev->ie_id.id_epoch : 0,
+			prior_punch ? prior_punch->ie_id.id_epoch : 0);
+		if (entry->ie_id.id_epoch > epr->epr_hi)
+			break; /* Done */
+		if (entry->ie_id.id_epoch < epr->epr_lo) {
+			if (entry->ie_id.id_epoch <= punched) {
+				/* Skip entries outside of the range and
+				 * punched by the parent
+				 */
+				continue;
+			}
+			if (entry->ie_punch) {
+				/* Just save the prior punch entry */
+				prior_punch = entry;
+			} else {
+				/* A create covers the prior punch */
+				prior_punch = NULL;
+			}
+			goto next; /* skip to next entry */
+		}
+		if (discard || entry->ie_status == ILOG_REMOVED ||
+		    punched >= entry->ie_id.id_epoch) {
+			/* Remove stale entry or punched entry */
+			goto remove_entry;
+		}
+
+		if (prev != NULL && prev->ie_punch == entry->ie_punch) {
+			/* Remove redundant entry */
+			goto remove_entry;
+		}
+
+		if (!entry->ie_punch) {
+			/* Create is needed for now */
+			goto next;
+		}
+
+		if (prev == NULL)
+			goto remove_entry;
+
+		if (prev->ie_id.id_epoch < epr->epr_lo) {
+			/** Data punched is not in range */
+			prior_punch = entry;
+			goto next;
+		}
+
+		/* remove the punched create entry */
+		D_ASSERT(!prev->ie_punch);
+		rc = remove_ilog_entry(lctx, &toh, prev,
+				       &removed);
+		if (rc != 0)
+			goto done;
+
+		/* Punch is redundant or covers nothing.  Remove it. */
```

Comment:
I'm bit confusing here, how can we know this punch is redundant or covers nothing? It may cover data outside of current epr, right?

Reply:
in this case, the previous entry is a punch or a covered update